### PR TITLE
Chapter 2 completed!

### PR DESCRIPTION
**Pull request by RajPathare**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,7 +336,10 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
+subList x y z = 
+    if x >=0 && y >=x
+    then take (y - x + 1) (drop x z)
+    else []
```
Comment (chshersh) [inline src/Chapter2.hs:342]: Your implementation is correct when the right bound is lower than the left. But I think you can implement it more efficiently by adding a separate case to your check :slightly_smiling_face:

File: src/Chapter2.hs [hunk 2]
```diff
@@ -348,8 +351,10 @@ Implement a function that returns only the first half of a given list.
 >>> firstHalf "bca"
 "b"
 -}
--- PUT THE FUNCTION TYPE IN HERE
-firstHalf l = error "firstHalf: Not implemented!"
+firstHalf :: [a] -> [a]
+firstHalf l = 
+    let a = length l `div` 2 - 1
+    in subList 0 a l
```
Comment (chshersh) [inline src/Chapter2.hs:357]: Nice usage of `subList` here! Though, `subList` does more than required, and here it's enough to _take_ the first half of the list.

File: src/Chapter2.hs [hunk 3]
```diff
@@ -604,9 +612,10 @@ Implement a function that duplicates each element of the list
 "aabbaacc"
 
 -}
-duplicate :: [a] -> [a]
-duplicate = error "duplicate: Not implemented!"
 
+duplicate :: [a] -> [a]
+duplicate [] = []
+duplicate (x:y) = x : x : duplicate y
```
Comment (chshersh) [inline src/Chapter2.hs:618]: Idiomatically, the list tail is called _xs_ usually when the variable is `x`
```suggestion
duplicate (x:xs) = x : x : duplicate xs
```

File: src/Chapter2.hs [hunk 4]
```diff
@@ -620,7 +629,11 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven [x] = [x]
+takeEven (x:_:y) = x : takeEven y
```
Comment (chshersh) [inline src/Chapter2.hs:636]: Great pattern-matching skills! :trophy:

File: src/Chapter2.hs [hunk 5]
```diff
@@ -726,8 +739,10 @@ value of the element itself
 
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
+
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate [] = []
+smartReplicate l = concat (map (\x -> replicate x x) l)
```
Comment (chshersh) [inline src/Chapter2.hs:745]: Great! 👍🏼 
If you want to write this function even more elegantly, you can use the standard function `concatMap` which is a combination of `concat` and `map` as the name suggests 🙂
After that, you could notice, that you would be able to eta-reduce on the last argument :)

Also, a separate case with `[]` is redundant. `map` and `concat` already handle empty lists without problems :slightly_smiling_face:

File: src/Chapter2.hs [hunk 6]
```diff
@@ -857,7 +883,11 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+
+rewind :: [a] -> [a]
+rewind [] = []
+rewind [x] = [x]
+rewind (x:y) = rewind y ++ [x]
```
Comment (chshersh) [inline src/Chapter2.hs:890]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by chshersh — APPROVED** [review]
Nice solutions for Chapter Two tasks! And well done with the Advanced Task :clap: 

I left some comments and suggestions. Your solutions are great, but you can make them even more awesome :100:

**Comment by vrom911** [comment]
I'm going to close this PR, because we can't merge the solutions to our repo :slightly_smiling_face: 
Feel free to open PRs with solutions to other chapters!

**Pull request closed without merging**